feat(website): document customization provider and theming (#1536)
twilio-labs/paste: packages/paste-website/src/pages/customization/customization-provider.mdx

        description: PackageJSON.description,
        version: PackageJSON.version,
      },
    },
  ]}
/>

---

<contentwrapper>

<PageAside data={props.data.mdx} />

<content>

The `CustomizationProvider` provides a central place for customizations to be applied to an application that is using Paste.

Say you want to build a web application or website using Paste, but want to have your own look and feel—the `CustomizationProvider` exists to do just that.

If you&rsquo;re a Flex customer, and want to build a Flex plugin that follows your brand guidelines, the `CustomizationProvider` can be used for that as well.

<ProgrammableAppsCallout />

## About the Customization Provider

Similar to the Paste [ThemeProvider](/tokens/theme-package), the `CustomizationProvider` uses [React Context](https://reactjs.org/docs/context.html) to provide a [theme object](/tokens/theme-package/#theme) to any descendant components in the tree.

The `CustomizationProvider` currently differs in two ways:

1. Firstly, you set a `baseTheme`. The base theme tells the provider which Paste theme you would like to base your custom theme on. You can choose from the 'Default' or 'Dark' Theme.

> 2. Secondly, you set a `theme`, but unlike with the `<Theme.Provider />` you pass it an object.

Reviewer: Have you considered keeping `theme` the same and adding a `customizations` prop instead?

Author: I hadn't entirely, but it was starting to bug me. I think it should be theme related as that's what it is, and `customizations` might not be explicit enough. But as this prop is a partial and merged into the base theme, `theme` is also not telling the the full story

`customTheme`? `themeOverrides`?

Reviewer: `overrides` sounds good

Author: Another data point. Both [material](https://material-ui.com/customization/theming/) and [chakra](https://chakra-ui.com/docs/theming/customize-theme) use `theme` but provide a helper function to merge overrides into a base theme. 

I wonder if there's any merit in keeping that API similar for transferability? Is there value in a helper for creating a theme and not overloading the provider

```
import {CustomizationProvider} from '@twilio-paste/core/customization';

const MyApp = () => (
  <CustomizationProvider baseTheme="Dark" theme={{
    "backgroundColors": {
      "colorBackgroundPrimary": "rgb(80, 123, 30)",
      "colorBackgroundPrimaryDarker": "rgb(56, 86, 21)",
      "colorBackgroundPrimaryLighter": "rgb(154,199,232)",
      "colorBackgroundPrimaryLightest": "rgb(197, 223, 242)",
    }
  }}>
    rest of app
  </CustomizationProvider>
)
```

This will output the Paste theme of your choice with the custom background color tokens above. So any component using `colorBackgroundPrimary`, `colorBackgroundPrimaryDarker`, etc will use the custom background color token values instead of the background color token values provided by the base theme.

The theme object should follow the following format, where each theme key can take any valid [design token](/tokens) for that grouping:

<ThemeObjectDisplay />

> Custom themes can also be imported as a json object, along with the `CustomizationProvider`. Remember, the custom theme must follow the format of the Paste theme object.

Reviewer: I don't really understand what this bit is saying. Specifically the `, along with the CustomizationProvider.` part

Author: Fair, maybe I can just reword it slightly. It's trying to say you need to import the json object, much in the same way you import the customization provider at the top of your file. But that might not need saying.

Author: This might be better

> After importing the custom theme needs to be passed to the `theme` prop on the `CustomizationProvider`. The `CustomizationProvider` should wrap the root of your application, or part of the application you would like to theme.

Reviewer: ```suggestion
The `CustomizationProvider` should wrap the root of your application, or a part of the application you would like to theme.
```
